### ISSUE TITLE
Names of bound variables of existing lemma in section change depending on local arguments

### ISSUE THREAD

**Issue by Zimmi48** [Zimmi48/bugzilla-test#4536]
Note: the issue was created automatically with bugzilla2github tool

# Bugzilla Bug ID: 4536
Date: 2016-01-26 21:32:24 +0100
From: Ralf Jung &lt;<post+coq@ralfj.de>&gt;
To:   
CC:   coq-bugs-redist@lists.gforge.inria.fr, @_silene, @_herbelin

Last updated: 2016-05-04 15:20:47 +0200

## Bugzilla Comment ID: 11720
Date: 2016-01-26 21:32:24 +0100
From: Ralf Jung &lt;<post+coq@ralfj.de>&gt;

I will let the code speak:

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
Section S.
  Context {n: nat}.
  Lemma lem1 (P Q : Prop) : n = O -> Q -> P -> Q.
  Proof. tauto. Qed.
End S.

Lemma lem2 (P' Q: Prop) : Q -> P' -> Q.
Proof.
  (* As expected, this eapply works: I am manually specifying what P should be. *)
  intros ? ?. eapply lem1 with (P := P').
Abort.

Lemma lem3 (P Q: Prop) : Q -> P -> Q.
Proof.
  (* Very surprisingly, this eapply does *not* work. Coq claims that there is no bound variable P,
     there is just P0. It seems that the names of the bound variables of lem1 actually *change*
     just because lem3 has itself an argument called P. *)
  intros ? ?. eapply lem1 with (P := P).
<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<

This doesn't make any sense, the names of bound variables of other lemmas shouldn't just change. To make this even more confusing, this bug only happens if lem1 was defined in a Section and actually uses one of the section parameters. (That's why I added that "n: nat".)

Consider me confused...

## Bugzilla Comment ID: 12396
Date: 2016-05-04 14:10:17 +0200
From: @_silene

I am in full agreement. That being said, the code goes to great lengths to achieve it. (There is even a function called "rename_bound_vars_as_displayed" just for that purpose.) So I very much doubt it will ever be changed to some saner semantics.

## Bugzilla Comment ID: 12397
Date: 2016-05-04 14:15:26 +0200
From: Ralf Jung &lt;<post+coq@ralfj.de>&gt;

Wait, what?? This is horrible! Please reopen.

This bug completely kills any kind of automation that wants to apply interesting lemmas -- you cannot rely on how the variables are called, so you cannot give "later" implicit arguments without also giving "earlier" ones. It is surprising, unpredictable (as it depends on whether the lemma was defined in a section), in other words -- plainly broken.

Did you find any trace in the good of something like a *reason* for doing strange things like this?

## Bugzilla Comment ID: 12398
Date: 2016-05-04 14:18:02 +0200
From: @_silene

Please understand that the only reason I am closing this bug-report is that I believe it will stay open for years, never ever getting any consideration. (I have handled enough bug-reports to know how it works, unfortunately.) I suggest you move the discussion to the mailing-list in order to get support from sufficiently many developers.

## Bugzilla Comment ID: 12399
Date: 2016-05-04 14:27:54 +0200
From: @_maximedenes

I had been trying to reproduce this bug for a while, because it occurred in VST. But I was missing the dependency on section variables. So thanks!

This indeed looks horrible, so I think we should discuss it. However, in my personal experience, you can often find a good ordering for implicit arguments that removes the need for (X := ...) binders which are fragile anyway.

## Bugzilla Comment ID: 12400
Date: 2016-05-04 14:47:01 +0200
From: Ralf Jung &lt;<post+coq@ralfj.de>&gt;

> This indeed looks horrible, so I think we should discuss it.

I sent a mail to the list.

> the need for (X := ...) binders which are fragile anyway.

Why are they fragile? I actually think they are *less* fragile than "foo _ _ _ _ _ a _ _ b". At least, if their behavior would be anything close to sane, they could be used to make code less subject to breakage as arguments are reordered or additional arguments are added in other parts of the codebase.

## Bugzilla Comment ID: 12401
Date: 2016-05-04 15:15:01 +0200
From: @_herbelin

No need for "horrible"s... here is the explanation:

Consider "apply (@lem1 (let a:=P in 0))". How would you like the bound variable P be displayed?

Consider "apply (@lem1 0)". How would you like the bound variable P be displayed?

Consider "apply @lem1". How would you like the bound variable P be displayed?

In short, doing "Set Printing Existential Instances", one sees that

Check lem1.

gives

lem1
     : forall P0 Q0 : Prop,
       ?n@{P:=P; Q:=Q; H:=H; H0:=H0} = 0 -> Q0 -> P0 -> Q0

which gives no other choices than renaming P since otherwise it would bind the P which occurs in the instance of the existential variable ?n.

So, I believe the misunderstanding comes from the fact that lem1 has an invisible maximal implicit argument which forces a change of binding so that bindings remain correct.
